In [1]:
import pandas as pd
import numpy as np


In [3]:
temp_df = pd.read_csv('IMDB Dataset.csv')

In [4]:
temp_df.shape

(50000, 2)

In [5]:
df = temp_df[:10000]

In [6]:
df.shape

(10000, 2)

In [7]:
df.head()

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [9]:
df.duplicated().value_counts()

False    9983
True       17
Name: count, dtype: int64

In [10]:
df.drop_duplicates(inplace =True)

C:\Users\User\AppData\Local\Temp\ipykernel_18300\1154324544.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop_duplicates(inplace =True)


In [11]:
df.shape

(9983, 2)

# Remove Punctions mark 

In [12]:
import re
def remove_tag(raw_text):
    cleaned_text = re.sub(re.compile('<.*?>') , '' ,raw_text)
    return cleaned_text

In [13]:
df['review'] = df['review'].apply(remove_tag)

C:\Users\User\AppData\Local\Temp\ipykernel_18300\156910529.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['review'] = df['review'].apply(remove_tag)


In [14]:
df['review']

0       One of the other reviewers has mentioned that ...
1       A wonderful little production. The filming tec...
2       I thought this was a wonderful way to spend ti...
3       Basically there's a family where a little boy ...
4       Petter Mattei's "Love in the Time of Money" is...
                              ...                        
9995    Fun, entertaining movie about WWII German spy ...
9996    Give me a break. How can anyone say that this ...
9997    This movie is a bad movie. But after watching ...
9998    This is a movie that was probably made to ente...
9999    Smashing film about film-making. Shows the int...
Name: review, Length: 9983, dtype: object

# Covert into Lower case 

In [15]:
df['review'] =df['review'].apply(lambda x : x.lower())


C:\Users\User\AppData\Local\Temp\ipykernel_18300\768984611.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['review'] =df['review'].apply(lambda x : x.lower())


In [16]:
df['review'] 

0       one of the other reviewers has mentioned that ...
1       a wonderful little production. the filming tec...
2       i thought this was a wonderful way to spend ti...
3       basically there's a family where a little boy ...
4       petter mattei's "love in the time of money" is...
                              ...                        
9995    fun, entertaining movie about wwii german spy ...
9996    give me a break. how can anyone say that this ...
9997    this movie is a bad movie. but after watching ...
9998    this is a movie that was probably made to ente...
9999    smashing film about film-making. shows the int...
Name: review, Length: 9983, dtype: object

# Remove the stopwords

In [18]:
pip install nltk

  Obtaining dependency information for nltk from https://files.pythonhosted.org/packages/60/90/81ac364ef94209c100e12579629dc92bf7a709a84af32f8c551b02c07e94/nltk-3.9.2-py3-none-any.whl.metadata
  Obtaining dependency information for click from https://files.pythonhosted.org/packages/98/78/01c019cdb5d6498122777c1a43056ebb3ebfeef2076d9d026bfe15583b2b/click-8.3.1-py3-none-any.whl.metadata
  Obtaining dependency information for joblib from https://files.pythonhosted.org/packages/7b/91/984aca2ec129e2757d1e4e3c81c3fcda9d0f85b74670a094cc443d9ee949/joblib-1.5.3-py3-none-any.whl.metadata
  Obtaining dependency information for regex>=2021.8.3 from https://files.pythonhosted.org/packages/f2/36/39d0b29d087e2b11fd8191e15e81cce1b635fcc845297c67f11d0d19274d/regex-2026.1.15-cp312-cp312-win_amd64.whl.metadata
     ---------------------------------------- 0.0/41.5 kB ? eta -:--:--
     ---------------------------------------- 0.0/41.5 kB ? eta -:--:--
     --------- ------------------------------ 10.2/41


[notice] A new release of pip is available: 23.2.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [19]:
from nltk.corpus import stopwords

sw_list = stopwords.words('english')

In [20]:
sw_list

['a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 "he'd",
 "he'll",
 'her',
 'here',
 'hers',
 'herself',
 "he's",
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 'if',
 "i'll",
 "i'm",
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 "i've",
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'on

In [21]:
df['review'] = df['review'].apply(lambda x : [item for item in x.split() if item not in sw_list]).apply(lambda x : " ".join(x))

C:\Users\User\AppData\Local\Temp\ipykernel_18300\4135132303.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['review'] = df['review'].apply(lambda x : [item for item in x.split() if item not in sw_list]).apply(lambda x : " ".join(x))


In [22]:
df['review']

0       one reviewers mentioned watching 1 oz episode ...
1       wonderful little production. filming technique...
2       thought wonderful way spend time hot summer we...
3       basically there's family little boy (jake) thi...
4       petter mattei's "love time money" visually stu...
                              ...                        
9995    fun, entertaining movie wwii german spy (julie...
9996    give break. anyone say "good hockey movie"? kn...
9997    movie bad movie. watching endless series bad h...
9998    movie probably made entertain middle school, e...
9999    smashing film film-making. shows intense stran...
Name: review, Length: 9983, dtype: object

In [24]:
pip install gensim

  Obtaining dependency information for gensim from https://files.pythonhosted.org/packages/fd/f2/9ec6863143888bf390cdc5261f6d9e71d79bc95d98fb815679dba478d5f6/gensim-4.4.0-cp312-cp312-win_amd64.whl.metadata
  Obtaining dependency information for scipy>=1.7.0 from https://files.pythonhosted.org/packages/92/ce/672ed546f96d5d41ae78c4b9b02006cedd0b3d6f2bf5bb76ea455c320c28/scipy-1.17.0-cp312-cp312-win_amd64.whl.metadata
     ---------------------------------------- 0.0/61.0 kB ? eta -:--:--
     ---------------------------------------- 0.0/61.0 kB ? eta -:--:--
     ------ --------------------------------- 10.2/61.0 kB ? eta -:--:--
     ------------------------- ------------ 41.0/61.0 kB 393.8 kB/s eta 0:00:01
     -------------------------------------- 61.0/61.0 kB 539.5 kB/s eta 0:00:00
  Obtaining dependency information for smart_open>=1.8.1 from https://files.pythonhosted.org/packages/ad/95/bc978be7ea0babf2fb48a414b6afaad414c6a9e8b1eafc5b8a53c030381a/smart_open-7.5.0-py3-none-any.whl.me


[notice] A new release of pip is available: 23.2.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


## 1) প্রয়োজনীয় লাইব্রেরি ইমপোর্ট

```python
import gensim
from nltk import sent_tokenize
from gensim.utils import simple_preprocess
```

* **gensim**: Word2Vec ট্রেন করার জন্য
* **sent_tokenize**: একটা বড় review কে বাক্যে ভাগ করবে (sentence segmentation)
* **simple_preprocess**: বাক্যকে ছোট করে টোকেন বানাবে

  * lowercase করবে
  * punctuation বাদ দেবে
  * শব্দগুলোর লিস্ট বানাবে

---

## 2) Training data বানানো (story লিস্ট)

```python
story = []
for doc in df['review']:
    raw_sent = sent_tokenize(doc)
    for sent in raw_sent:
        story.append(simple_preprocess(sent))
```

এখানে কী হচ্ছে:

* `df['review']` → প্রতিটা রিভিউ (`doc`) নিয়ে কাজ
* `sent_tokenize(doc)` → রিভিউকে sentence list বানায়
  উদাহরণ: `"Good phone. Battery is great!"` → `["Good phone.", "Battery is great!"]`
* তারপর প্রতিটা sentence কে `simple_preprocess(sent)` দিয়ে token list বানায়
  `"Battery is great!"` → `["battery", "is", "great"]`
* এই token list গুলো `story` তে জমা হয়

**ফাইনাল `story` কেমন?**
`story` হলো `list of list`:

```text
[
  ["good", "phone"],
  ["battery", "is", "great"],
  ...
]
```

Word2Vec ট্রেন করার জন্য ঠিক এই ফরম্যাটটাই লাগে।

---

## 3) Word2Vec মডেল তৈরি

```python
model = gensim.models.Word2Vec(
    window=10,
    min_count=2
)
```

এখানে দুইটা গুরুত্বপূর্ণ সেটিং:

* `window=10`
  একটি শব্দের আশেপাশে 10টা শব্দ পর্যন্ত দেখে “context” শিখবে।
  window বড় হলে বেশি context, ছোট হলে কম context।

* `min_count=2`
  যেসব শব্দ পুরো কর্পাসে **২ বারও আসে না**, সেগুলো বাদ।
  কারণ খুব rare শব্দ noise বাড়ায়।

> নোট: এখানে আপনি `vector_size`, `sg`, `workers` ইত্যাদি দেননি—তাই default value ব্যবহার হবে।

---

## 4) Vocabulary তৈরি + Training

```python
model.build_vocab(story)
model.train(story, total_examples=model.corpus_count, epochs=model.epochs)
(5875535, 6212140)
```

এটা দুই ধাপ:

### (ক) `build_vocab(story)`

* `story` থেকে সব unique শব্দ বের করে vocabulary বানায়
* তারপর যেগুলো `min_count` কম, সেগুলো বাদ দেয়

### (খ) `train(...)`

* Word2Vec শিখে: কোন শব্দ কোন শব্দের সাথে সাধারণত আসে
* আউটপুট `(5875535, 6212140)` টাইপ কিছু দেখালে সাধারণত বোঝায়:

  * ট্রেনিংয়ে কত words/process হয়েছে (প্রায়)
  * আর কত total words ছিল (প্রায়)

---

## 5) Vocabulary size দেখা

```python
len(model.wv.index_to_key)
31845
```

* `model.wv.index_to_key` = মডেল যেসব শব্দ জানে (vocabulary)
* এখানে মডেল **31845** টি শব্দ শিখেছে

---

## 6) Document vector (একটা রিভিউকে vector বানানো)

```python
def document_vector(doc):
    # remove out-of-vocabulary words
    doc = [word for word in doc.split() if word in model.wv.index_to_key]
    return np.mean(model.wv[doc], axis=0)
```

এটা কোডের সবচেয়ে গুরুত্বপূর্ণ অংশ।

### ভেঙে বলি:

#### (ক) `doc.split()`

* রিভিউ টেক্সটকে স্পেস দিয়ে ভাগ করে শব্দ বানায়
  `"this phone is good"` → `["this","phone","is","good"]`

#### (খ) OOV বাদ দেওয়া

```python
[word for word in doc.split() if word in model.wv.index_to_key]
```

* যেসব শব্দ Word2Vec vocabulary-তে নেই (Out-Of-Vocabulary), সেগুলো বাদ
* কারণ মডেল যেটা শেখেনি সেটার vector বের করা যাবে না

#### (গ) প্রতিটা শব্দের vector বের করা

```python
model.wv[doc]
```

* `doc` এখন শব্দের লিস্ট
* `model.wv[doc]` → shape হবে:
  `(num_words_in_doc, vector_size)`

#### (ঘ) Mean নিয়ে document embedding বানানো

```python
np.mean(..., axis=0)
```

* প্রতিটা শব্দের vector এর গড় নিয়ে **একটা fixed vector** বানানো হয়
* ধরুন vector_size = 100 হলে output হবে 100-dimension vector

আপনার আউটপুট:

```python
array([...], dtype=float32)
```

এটা ওই রিভিউর embedding।

> সহজভাবে: **Word2Vec শব্দকে সংখ্যা বানায়, আর আপনি শব্দগুলোর গড় নিয়ে রিভিউকে সংখ্যা বানাচ্ছেন।**

---

## 7) সব রিভিউর জন্য X বানানো

```python
from tqdm import tqdm
X = []
for doc in tqdm(df['review'].values):
    X.append(document_vector(doc))

X = np.array(X)
```

* `tqdm` শুধু progress bar দেখায়
* `X` হবে feature matrix:

  * shape: `(num_reviews, vector_size)`
* এখানে আপনার `df` তে 9983টা রিভিউ, তাই:

  * `X.shape` হবে প্রায় `(9983, 100)` (vector_size যদি 100 হয়)

---

## 8) Label encoding (sentiment কে 0/1 বানানো)

```python
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()

y = encoder.fit_transform(df['sentiment'])
```

* `df['sentiment']` যদি `"positive"/"negative"` টাইপ string থাকে:

  * সেগুলোকে numeric বানায়
* আপনার আউটপুটে দেখা যাচ্ছে:

  * `array([1, 1, 1, ..., 0, 0, 1])`

মানে:

* একটা sentiment class = 1
* আরেকটা sentiment class = 0

> কোনটা 0 আর কোনটা 1 হয়েছে—এটা জানতে `encoder.classes_` দেখতে হয়।
> (আপনি কোডে দেখাননি, কিন্তু সেটা গুরুত্বপূর্ণ।)

---

## 9) Train-test split

```python
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=1)
```

* 80% train, 20% test
* `random_state=1` দিলে প্রতিবার একইভাবে split হবে (reproducible)

---

## 10) RandomForest দিয়ে ক্লাসিফিকেশন

```python
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

rf = RandomForestClassifier()
rf.fit(X_train,y_train)
y_pred = rf.predict(X_test)
accuracy_score(y_test,y_pred)
```

* `RandomForestClassifier()` default সেটিংয়ে অনেকগুলো decision tree তৈরি করে
* `fit()` → train data দিয়ে শেখে
* `predict()` → test set এ লেবেল প্রেডিক্ট করে
* `accuracy_score()` → কত শতাংশ ঠিক বলেছে তা দেয়

---

# আপনার পাইপলাইনটা এক লাইনে সারাংশ

**Text → Sentence tokenize → Word tokenize → Word2Vec train → Review vector (mean of word vectors) → RandomForest → Accuracy**


In [25]:
import gensim

In [26]:
from nltk import sent_tokenize
from gensim.utils import simple_preprocess

In [42]:
story = []

for doc in df['review']:
    raw_sent = sent_tokenize(doc)
    for sent in raw_sent:
        story.append(simple_preprocess)

In [44]:
story

[<function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_len=2, max_len=15)>,
 <function gensim.utils.simple_preprocess(doc, deacc=False, min_l

In [45]:
model = gensim.models.Word2Vec(
    window = 10,
    min_count = 2
)

In [46]:
model.build_vocab(story)

TypeError: 'function' object is not iterable

In [48]:
from nltk import sent_tokenize
from gensim.utils import simple_preprocess

story = []
for doc in df["review"].astype(str).values:
    for sent in sent_tokenize(doc):
        tokens = simple_preprocess(sent)
        if tokens:
            story.append(tokens)

print(type(story), type(story[0]), story[0][:10])  # sanity check


<class 'list'> <class 'list'> ['one', 'reviewers', 'mentioned', 'watching', 'oz', 'episode', 'hooked']
